Source/Core/Core/PowerPC/PPCCache.cpp

```diff
@@ -93,63 +93,139 @@ InstructionCache::~InstructionCache()
+void Cache::Store(u32 addr)
 {
-  if (!HID0.ICE || m_disable_icache)
+  u32 set;
+  u32 t;
+  GetCache(addr, set, t, true);
+
+  if (t == 0xff)
     return;
 
-  // Invalidates the whole set
-  const u32 set = (addr >> 5) & 0x7f;
-  for (size_t i = 0; i < 8; i++)
+  if (valid[set] & (1U << t) && wrote[set] & (1U << t))
+    Memory::CopyToEmu((addr & ~0x1f), reinterpret_cast<u8*>(data[set][t].data()), 32);
+  wrote[set] &= ~(1U << t);
+}
+
+void Cache::FlushAll()
+{
+  for (size_t set = 0; set < CACHE_SETS; set++)
   {
-    if (valid[set] & (1U << i))
+    for (size_t way = 0; way < CACHE_WAYS; way++)
     {
-      if (tags[set][i] & (ICACHE_VMEM_BIT >> 12))
-        lookup_table_vmem[((tags[set][i] << 7) | set) & 0xfffff] = 0xff;
-      else if (tags[set][i] & (ICACHE_EXRAM_BIT >> 12))
-        lookup_table_ex[((tags[set][i] << 7) | set) & 0x1fffff] = 0xff;
-      else
-        lookup_table[((tags[set][i] << 7) | set) & 0xfffff] = 0xff;
+      if (valid[set] & (1U << way) && wrote[set] & (1U << way))
+        Memory::CopyToEmu(addrs[set][way], reinterpret_cast<u8*>(data[set][way].data()), 32);
     }
   }
-  valid[set] = 0;
-  JitInterface::InvalidateICacheLine(addr);
+
+  Reset();
 }
 
-u32 InstructionCache::ReadInstruction(u32 addr)
+void Cache::Invalidate(u32 addr)
 {
-  if (!HID0.ICE || m_disable_icache)  // instruction cache is disabled
-    return Memory::Read_U32(addr);
-  u32 set = (addr >> 5) & 0x7f;
-  u32 tag = addr >> 12;
+  u32 set;
+  u32 t;
+  GetCache(addr, set, t, true);
+
+  if (t == 0xff)
+    return;
+
+  if (valid[set] & (1U << t))
+  {
+    if (tags[set][t] & (CACHE_VMEM_BIT >> 12))
+      lookup_table_vmem[((tags[set][t] << 7) | set) & 0xfffff] = 0xff;
+    else if (tags[set][t] & (CACHE_EXRAM_BIT >> 12))
+      lookup_table_ex[((tags[set][t] << 7) | set) & 0x1fffff] = 0xff;
+    else
+      lookup_table[((tags[set][t] << 7) | set) & 0xfffff] = 0xff;
+
+    valid[set] &= ~(1U << t);
+    wrote[set] &= ~(1U << t);
+  }
+}
+
+void Cache::Flush(u32 addr)
+{
+  u32 set;
+  u32 t;
+  GetCache(addr, set, t, true);
+
+  if (t == 0xff)
+    return;
+
+  if (valid[set] & (1U << t))
+  {
+    if (wrote[set] & (1U << t))
+      Memory::CopyToEmu((addr & ~0x1f), reinterpret_cast<u8*>(data[set][t].data()), 32);
+
+    if (tags[set][t] & (CACHE_VMEM_BIT >> 12))
+      lookup_table_vmem[((tags[set][t] << 7) | set) & 0xfffff] = 0xff;
+    else if (tags[set][t] & (CACHE_EXRAM_BIT >> 12))
+      lookup_table_ex[((tags[set][t] << 7) | set) & 0x1fffff] = 0xff;
+    else
+      lookup_table[((tags[set][t] << 7) | set) & 0xfffff] = 0xff;
 
+    valid[set] &= ~(1U << t);
+    wrote[set] &= ~(1U << t);
+  }
+}
+
+void Cache::Touch(u32 addr, bool store)
+{
+  u32 set;
   u32 t;
-  if (addr & ICACHE_VMEM_BIT)
+  GetCache(addr, set, t, true);
+
+  if (t == 0xff)
+    return;
+
+  if (store && valid[set] & (1U << t))
+    wrote[set] |= (1U << t);
+}
+
+void Cache::GetCache(u32 addr, u32& set, u32& t, bool locked)
```

Comment:
Perhaps return the `set` and `t` values as a `std::pair` or a named struct and then make usages of the function like:

```cpp
auto [set, t] = GetCache(...);
```

So that it gets rid of the need for out parameters, and also completely eliminates any potential for uninitialized reads/writes being unintentionally introduced through refactoring in the future.